libtenzir/include/tenzir/data_builder.hpp

```diff
@@ -0,0 +1,575 @@
+  old_index = type_index_generic_mismatch;
+}
+
+enum class state { alive, sentinel, dead };
+
+class node_base {
+protected:
+  auto mark_this_relevant() -> void {
+    if (state_ != state::alive) {
+      state_ = state::sentinel;
+    }
+  }
+  auto mark_this_dead() -> void {
+    state_ = state::dead;
+  }
+  auto mark_this_alive() -> void {
+    state_ = state::alive;
+  }
+  auto is_dead() const -> bool;
+  auto is_alive() const -> bool;
+  auto affects_signature() const -> bool;
+  state state_ = state::alive;
+};
+
+class node_record : public node_base {
+  friend class node_list;
+  friend class node_object;
+  friend class ::tenzir::data_builder;
+
+public:
+  /// @brief Reserves storage for at least N elements in the record.
+  /// this function can be used to get temporary pointer stability on the
+  /// records elements
+  auto reserve(size_t N) -> void;
+  /// @brief Adds a field to the record.
+  /// @note the returned pointer is not permanently stable. If the underlying
+  /// vector reallocates, the pointer becomes invalid
+  /// @ref reserve can be used to ensure stability for a given number of elements
+  [[nodiscard]] auto field(std::string_view name) -> node_object*;
+
+private:
+  // tries to get a field with the given name. Does not affect any field state
+  auto try_field(std::string_view name) -> node_object*;
+  // does lookup of a (nested( key
+  auto at(std::string_view key) -> node_object*;
+  // writes the record into a series builder
+  auto
+  commit_to(tenzir::record_ref r, class data_builder& rb,
+            const tenzir::record_type* seed, bool mark_dead = true) -> void;
+  auto
+  commit_to(tenzir::record& r, class data_builder& rb,
+            const tenzir::record_type* seed, bool mark_dead = true) -> void;
+  // append the signature of this record to `sig`.
+  // including sentinels is important for signature computation
+  auto append_to_signature(signature_type& sig, class data_builder& rb,
+                           const tenzir::record_type* seed) -> void;
+  // clears the record by marking everything as dead
+  auto clear() -> void;
+
+  // Record entry. This contains a string for the key and a field.
+  // Its defined out of line because node_object cannot be defined at this point.
+  struct entry_type;
+  // This stores added fields in order of their appearance
+  // This order is used for committing to the `series_builder`, in order to
+  // (mostly) preserved the field order from the input, apart from fields the
+  // `series_builder` was seeded with. The order of fields in a seed/selector on
+  // the other hand is then practically ensured because the multi_series_builder
+  // first seeds the respective `series_builder`.
+  std::vector<entry_type> data_;
+  // This is a sorted key -> index map. It is used for signature computation.
+  // If this map is not sorted, the signature computation algorithm breaks,
+  // since it would then be order dependent.
+  flat_map<std::string, size_t> lookup_;
```

Comment:
Thanks, but if the input order is inconsistent and a record is replaced, then the resulting builder will also use the newly derived order when adding the fields to the inner builder, which means that you will eventually get outputs of a different schema anyway, even though the signature stays the same. In a sense, the signature is even wrong for that case. So that cannot be the argument in its entirety.

I think you can adapt it though for the scenario where we have a schema (wherever it comes from). We want events of the same output schema to land in the same builder. However, seed fields are lazily appended just before computing the signature, so `data_` does not reflect the actual order in which the events will end up in the output, nor does it have a consistent ordering that is based on the input (as you mentioned). But the fields that belong to the schema must have a consistent order. The other fields do not really matter, because fields that are not in the schema will have varying order anyway (same argument as above). One way to put schema fields in consistent order is to sort _every_ field alphabetically (which does a bit more than just that). But we could also iterate over the schema fields first, and then do the remainder of the non-schema fields. You can still get a varying signature when a record is replaced, but only if the output schema is also varying.

I'm not saying that we have to do that nor something similar. But it's good to understand why doing nothing does not work. Perhaps you can add a summary of this case to your comment (or link my comment here if you want to). Maybe it helps future readers :+1:

Reply:
The order of fields in the schema is ensured by the `series_builder` though, is it not? It should have all fields it has been created with first, followed fields that were added later. Similarly, for the signature computation it should have "no effect", because in all cases where we have a seed, the fields are part of the signature in sorted order.

The series_builder resetting on on yield is a good point though. The only way that could be addressed would be by not resetting the builder (e.g. via some special flag)

Reply:
> The order of fields in the schema is ensured by the series_builder though, is it not?

Initializing `series_builder` with a schema will maintain the initialized fields and their order. This does not extend to fields that are added afterwards. `series_builder` was built with the assumption that you want (if possible) a single series that matches your input between flushes as close as possible. 

> It should have all fields it has been created with first, followed fields that were added later. 

Fields that are added later indeed come later, but once you flush the builder, they do not exist anymore. If you re-add them, their order depends completely on the order that they are inserted again afterwards.